examples/greeter-extension/src/greeter-extension-point.ts

```diff
@@ -0,0 +1,63 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/example-greeter-extension
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Getter} from '@loopback/context';
+import chalk from 'chalk';
+import {configuration, extensionPoint, extensions} from './decorators';
+import {Greeter} from './types';
+
+/**
+ * Options for the greeter extension point
+ */
+export interface GreeterExtensionPointOptions {
+  color: string;
+}
+
+/**
+ * An extension point for greeters that can greet in different languages
+ */
+@extensionPoint('greeter')
+export class GreeterExtensionPoint {
+  constructor(
+    /**
+     * Inject a getter function to fetch greeters (bindings tagged with
+     * 'greeter')
+     */
+    @extensions() // Sugar for @inject.getter(filterByTag({extensionPoint: 'greeter'}))
+    private greeters: Getter<Greeter[]>,
+    /**
+     * An extension point should be able to receive its options via dependency
+     * injection.
+     */
+    @configuration() // Sugar for @inject('greeter-extension-point.options', {optional: true})
+    private options?: GreeterExtensionPointOptions,
+  ) {}
+
+  /**
+   * Greet in the given language
+   * @param language Language code
+   * @param name Name
+   */
+  async greet(language: string, name: string): Promise<string> {
+    let greeting: string = '';
+    // Get the latest list of greeters
+    const greeters = await this.greeters();
```

Comment:
_Re-posting https://github.com/strongloop/loopback-next/pull/2249/files#r267735009:_

If we end up with a single class implementing both greeter lookup and invocation of the `greet`, then we should at least split the implementation of this method into two parts.

```ts
export class GreeterExtensionPoint {
  // ...

async greet(language: string, name: string): Promise<string> {
  const greeter = this.findGreeterByLanguage(language);

  const greeting = greeter
    ? greeter.greet(name):
    : `Hello, ${name}`;  // Fall back to English

  if (this.options && this.options.color) {
    greeting = chalk.keyword(this.options.color)(greeting);
  }
  return greeting;
}
```